[i18n] All files are translated in French language

thomasbnt opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://github.com/forem/forem/blob/main/CONTRIBUTING.md#create-a-pull-request.
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/main/CODE_OF_CONDUCT.md.
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.

     NOTE: Pull Requests from forked repositories will need to be reviewed by
     a Forem Team member before any CI builds will run. Once your PR is approved
     with a `/ci` reply to the PR, it will be allowed to run subsequent builds without
     manual approval.
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
i18n French translation for all files into locales folder. 

## Related Tickets & Documents

This discussion https://github.com/forem/forem/discussions/17107

<!--
For pull requests that relate or close an issue, please include them
below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "closes #1234" would connect the current pull
request to issue 1234.  And when we merge the pull request, Github will
automatically close the issue.
-->

- Related Issue #
- Closes #

## QA Instructions, Screenshots, Recordings

_Please replace this line with instructions on how to test your changes, a note
on the devices and browsers this has been tested on, as well as any relevant
images for UI changes._

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://developers.forem.com/frontend/accessibility)._

## Added/updated tests?

- [ ] Yes
- [x] No, and this is why: Translated words and phrases from english files.
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_


## What gif best describes this PR or how it makes you feel?

![Finally YES](https://tenor.com/bKeL9.gif)

djuber commented on config/locales/services/fr.yml:
```diff
@@ -2,111 +2,111 @@
     error_messages:
       clean:
-        parse_error: There was a problem parsing the front-matter YAML. Perhaps you need to escape a quote or a colon or something. Email %{email} if you are having trouble.
+        parse_error: Il y a eu un problème pour analyser la matière première YAML. Vous devez peut-être échapper un guillemet, un deux-points ou autre chose. Envoyez un courriel à %{email} si vous avez des problèmes.
```
Since the surrounding context is a yaml parse error, actual chevrons `«` would have been ignored as other text and not treated as punctuation.

https://fr.wikipedia.org/wiki/Guillemet#Guillemets_dactylographiques

I think I answered my own question, this is not something that would confuse a user, and _guillemet_ here includes single and double quotes.

djuber commented on config/locales/helpers/fr.yml:
```diff
@@ -4,231 +4,230 @@ fr:
+      any_of_those: n'importe lequel de ces éléments
+      email_password: Email et mot de passe
+      reminder:  'Rappel : vous avez utilisé %{method} pour authentifier votre compte, veuillez donc utiliser %{dem} pour vous connecter si vous y êtes invité'
+      that: que
```
check: "that" can either be used to qualify (_que_) or demonstrate (_ça / cela_) - check where we use this translation key for the right context.

Oh, it doesn't matter, looks like this key is used nowhere?

```
app/helpers/articles_helper.rb:4:      [I18n.t("helpers.articles_helper.recently_created"), "creation-desc"],
app/helpers/articles_helper.rb:5:      [I18n.t("helpers.articles_helper.recently_published"), "published-desc"],
app/helpers/articles_helper.rb:6:      [I18n.t("helpers.articles_helper.most_views"), "views-desc"],
app/helpers/articles_helper.rb:7:      [I18n.t("helpers.articles_helper.most_reactions"), "reactions-desc"],
app/helpers/articles_helper.rb:8:      [I18n.t("helpers.articles_helper.most_comments"), "comments-desc"],
app/helpers/articles_helper.rb:31:        title: I18n.t("helpers.articles_helper.logo", service_name: service_name),
```

thomasbnt replied:
I leave that for the moment like this?

djuber replied:
yes, from what I can see nothing to do here on your side.

- it doesn't matter how this "that" was interpreted since it's not used from what I see
- removing this is out of scope of the work you're doing here, and it's better to focus on one thing at a time

yheuhtozr replied:
FYI, the key is called here, as a demonstrative opposed to "any of those"; both to be injected as a part of the text of the key `reminder`.

https://github.com/forem/forem/blob/07fff2f8efe997477706d75a95ef0d25c12dc48e/app/helpers/authentication_helper.rb#L30

I'm not sure whether this portion of code is further called on or not, though.

djuber replied:
@yheuhtozr thanks for pointing this out, if it's a demonstrative then _ce_ rather than _que_ seems most appropriate here.

djuber commented on config/locales/helpers/fr.yml:
```diff
@@ -4,231 +4,230 @@ fr:
-        title: Limit number of updates to a post
-        description: How many updates can someone make within any 30 second period? Update API docs when changed.
+        title: Limiter le nombre de mises à jour d'un article
+        description: Combien de mises à jour peut-on faire sur une période de 30 secondes ? Mettre à jour la documentation de l'API en cas de modification.
```
oh, interesting! Since the [API docs](https://developers.forem.com/api) are for Forem generally, this hint to the administrator might not make any more sense for a site-configurable setting.

unrelated to your PR's changes, but since I'm reading lots of messages we show users in some context I'm taking notes of things that we might want to revisit.

djuber commented on config/locales/liquid_tags/fr.yml:
```diff
@@ -2,112 +2,112 @@
       github_readme_tag:
-        invalid_options: 'GitHub tag: invalid options: %{invalid} | supported options: %{valid}'
-        invalid_github_repository: Invalid GitHub repository path or URL
+        invalid_options: 'Balise GitHub : options invalides : %{invalid} | Options supportées : %{valid}'
```
note: this "GitHub tag" refers to the markup (liquid tag) used to embed github issues/repos, and not a "Tag" from other contexts (tags on posts or listings), [_balise_](https://fr.wikipedia.org/wiki/Langage_de_balisage) looks like it makes sense.

djuber reviewed (commented):
Thank you for this! It must have been a lot of work.

Left a few comments about possible inconsistency surrounding terms for email and tags. I'm an American English speaker, so this is equally attributable to unfamiliarity on my part with normal usage as it is to any correctable mistakes on your part.

Everything else was more reading notes. We have a number of keys in these locale files that don't appear to be used anywhere, and a translation for a feature we removed years ago, and finding this dead code and redundancies was worth the time spent reading through the files.

We have a few French speaking core team, who should have a chance to look at this before we approve.

thomasbnt commented on config/locales/controllers/fr.yml:
```diff
@@ -1,121 +1,121 @@
+    badge_notifications: notifications de badges
+    comment_notifications: notifications de commentaires
+    digest_emails: "%{community} courriels récapitulatifs"
+    follower_notifications: notifications de followers
```
In France, the term "Follower" is actually used quite a lot on social networks. 👍

thomasbnt commented on config/locales/helpers/fr.yml:
```diff
@@ -4,231 +4,230 @@ fr:
-        latest: Latest posts
+        week: Principaux messages de la semaine
+        month: Principaux messages du mois
+        year: Les meilleurs messages de l'année
```
Ooops. I must have missed this lousy translation.
But it's definitely "This week's best articles." (🇫🇷 : "Les meilleurs articles de cette semaine" )

thomasbnt commented on config/locales/controllers/fr.yml:
```diff
@@ -1,121 +1,121 @@
+    moderated: Non autorisé en raison d'une action de modération.
   confirmations_controller:
-    email_sent: Email sent! Please contact support at %{email} if you are having trouble receiving your confirmation instructions.
+    email_sent: Courriel envoyé ! Veuillez contacter le service d'assistance à l'adresse %{email} si vous avez des difficultés à recevoir les instructions de confirmation.
```
Good catch, it is better to use the term "E-mail/Email" instead of "courriel".

thomasbnt commented:
I adjusted some words and phrases @djuber, like `courriel` to `e-mail` for all french files, edited `guillemet` to `citation`, make more sense for me, but I would like someone to help me with this. An error with `Principaux messages` to `Les meilleurs articles`. 📚

djuber commented:
Looking at the failing tests (inconsistent interpolation, and more specifically the missing keys check), this mostly happens when there's a change in the yaml structure (key nesting/naming). I'll take a second look in the morning but this is often fixable by confirming indentation levels are consistent between the `en` and `fr` files (or, between the files in the branch and in main). 

```
        [{:key=>"fr.helpers.articles_helper.recently_published", :file=>"config/locales/helpers/fr.yml"}, 
{:key=>"fr.helpers.authentication_helper.email_password", :file=>"config/locales/helpers/fr.yml"}, 
{:key=>"fr.helpers.authentication_helper.any_of_those", :file=>"config/locales/helpers/fr.yml"}, 
{:key=>"fr.helpers.authentication_helper.that", :file=>"config/locales/helpers/fr.yml"}, 
{:key=>"fr.helpers.authentication_helper.reminder", :file=>"config/locales/helpers/fr.yml"}, 
{:key=>"fr.helpers.authentication_helper.invite_only", :file=>"config/locales/helpers/fr.yml"}, 
{:key=>"fr.models.profile_pin.only_five", :file=>"config/locales/models/fr.yml"}, {:key=>"fr.models.profile_pin.pin_unpermitted", 
:file=>"config/locales/models/fr.yml"}, {:key=>"fr.models.sponsorship.invalid_type", :file=>"config/locales/models/fr.yml"}, 
{:key=>"fr.models.sponsorship.already_sponsored", :file=>"config/locales/models/fr.yml"}, 
{:key=>"fr.models.sponsorship.only_one_level", :file=>"config/locales/models/fr.yml"}, {:key=>"fr.models.sponsorship.level.gold", 
:file=>"config/locales/models/fr.yml"}, {:key=>"fr.models.sponsorship.level.silver", :file=>"config/locales/models/fr.yml"}, 
{:key=>"fr.models.sponsorship.level.bronze", :file=>"config/locales/models/fr.yml"}, 
{:key=>"fr.datetime.expires_in.over_x_years.one", :file=>"config/locales/fr.yml"}, 
{:key=>"fr.datetime.expires_in.over_x_years.other", :file=>"config/locales/fr.yml"}, 
{:key=>"en.datetime.expires_in.over_x_yars.one", :file=>"config/locales/en.yml"}, 
{:key=>"en.datetime.expires_in.over_x_yars.other", :file=>"config/locales/en.yml"}, 
{:key=>"en.helpers.articles_helper.récemment_publié", :file=>"config/locales/helpers/en.yml"}, 
{:key=>"en.helpers.authentification_helper.any_of_those", :file=>"config/locales/helpers/en.yml"}, 
{:key=>"en.helpers.authentification_helper.email_password", :file=>"config/locales/helpers/en.yml"}, 
{:key=>"en.helpers.authentification_helper.reminder", :file=>"config/locales/helpers/en.yml"}, 
{:key=>"en.helpers.authentification_helper.that", :file=>"config/locales/helpers/en.yml"}, 
{:key=>"en.helpers.authentification_helper.invite_only", :file=>"config/locales/helpers/en.yml"}, 
{:key=>"en.models.profil_pin.only_five", :file=>"config/locales/models/en.yml"}, {:key=>"en.models.profil_pin.pin_unpermitted", 
:file=>"config/locales/models/en.yml"}, {:key=>"en.models.settings.sponsorship.invalid_type", :file=>"config/locales/models/en.yml"}, 
{:key=>"en.models.settings.sponsorship.already_sponsored", :file=>"config/locales/models/en.yml"}, 
{:key=>"en.models.settings.only_one_level", :file=>"config/locales/models/en.yml"}, {:key=>"en.models.settings.level.gold", 
:file=>"config/locales/models/en.yml"}, {:key=>"en.models.settings.level.silver", :file=>"config/locales/models/en.yml"}, 
{:key=>"en.models.settings.level.bronze", :file=>"config/locales/models/en.yml"}]
```

djuber commented:
It looks like the "inconsistent interpolations" showing as a test failure is due to translation of an argument (community -> communauté) in some of the translations:

```
djuber@forem:~/src/forem$ bundle exec i18n-tasks check-consistent-interpolations
Inconsistent interpolations (6) | i18n-tasks v1.0.9
+--------+---------------------------------------------------+------------------------------------------------------------------+
| Locale | Key                                               | Value                                                            |
+--------+---------------------------------------------------+------------------------------------------------------------------+
|   fr   | helpers.feedback_messages_helper.affected.subject | Avis de courtoisie de %{communauté}                              |
|   fr   | json.author                                       | Auteur actif de la %{communauté}                                 |
|   fr   | mailers.digest_mailer.more_top_posts              | plus de meilleurs articles de %{communtiy}                       |
|   fr   | mailers.digest_mailer.more_top_posts_based        | plus de meilleurs articles de %{communtiy} basé sur vos intérêts |
|   fr   | mailers.digest_mailer.other_top_posts             | d'autres articles tendance de la %{communauté}                   |
|   fr   | mailers.digest_mailer.other_trending_posts        | d'autres articles tendance de la %{communauté}.                  |
+--------+---------------------------------------------------+------------------------------------------------------------------+
```

Since these interpolated values are passed in from code/named keys, these should remain consistent with the English side (community) and not be translated.

djuber commented on config/locales/helpers/fr.yml:
```diff
@@ -4,230 +4,229 @@ fr:
+      most_reactions: Le plus de réactions
+      most_views: Le plus de vues
+      recently_created: Récemment créé
+      récemment_publié: Récemment publié
```
this key `récemment_publié` should be `recently_published` (to match the code calling the translation, this and the English key must be the same)

thomasbnt replied:
Oowww good catch !

djuber commented on config/locales/models/fr.yml:
```diff
@@ -2,131 +2,131 @@
-      only_five: cannot have more than five total pinned posts
-      pin_unpermitted: must have proper permissions for pin
+      cannot_vote_more_than_once: Impossible de voter plus d'une fois dans un sondage.
+    profil_pin:
```
similar to recently published, `profile_pin` is code (so untranslated)

djuber commented:
I think I see it - `authentification_helper` should have remained `authentication_helper` (in helpers/fr.yml). That probably explains the majority of the remaining missing keys.

djuber commented:
Was there a typo in the english file key that was fixed in the fr file? I assume we want to fix the english file (unless this `yars` typo is referenced elsewhere in code).

{:key=>"en.datetime.expires_in.over_x_yars.one", :file=>"config/locales/en.yml"}, 
{:key=>"en.datetime.expires_in.over_x_yars.other", :file=>"config/locales/en.yml"},

thomasbnt commented:
> I think I see it - `authentification_helper` should have remained `authentication_helper` (in helpers/fr.yml). That probably explains the majority of the remaining missing keys.


> `{:key=>"en.datetime.expires_in.over_x_yars.one", :file=>"config/locales/en.yml"},`
> 
> in the config/locales/fr.yml file, there's a key ` over_x_yars:` which doesn't match the English original (probably just an accidentally removed e)?

Both are edited ! Thanks again @djuber !

aitchiss commented:
It looks like we still have some spec failures:

```
Failures:
  1) I18n does not have missing keys
     Failure/Error:
       expect(missing_keys).to be_empty,
                               "Missing #{missing_keys.leaves.count} i18n keys\n #{details(missing_keys)}"
     
       Missing 10 i18n keys
        [{:key=>"fr.helpers.authentication_helper.email_password", :file=>"config/locales/helpers/fr.yml"}, {:key=>"fr.helpers.authentication_helper.any_of_those", :file=>"config/locales/helpers/fr.yml"}, {:key=>"fr.helpers.authentication_helper.that", :file=>"config/locales/helpers/fr.yml"}, {:key=>"fr.helpers.authentication_helper.reminder", :file=>"config/locales/helpers/fr.yml"}, {:key=>"fr.helpers.authentication_helper.invite_only", :file=>"config/locales/helpers/fr.yml"}, {:key=>"en.helpers.authentification_helper.any_of_those", :file=>"config/locales/helpers/en.yml"}, {:key=>"en.helpers.authentification_helper.email_password", :file=>"config/locales/helpers/en.yml"}, {:key=>"en.helpers.authentification_helper.reminder", :file=>"config/locales/helpers/en.yml"}, {:key=>"en.helpers.authentification_helper.that", :file=>"config/locales/helpers/en.yml"}, {:key=>"en.helpers.authentification_helper.invite_only", :file=>"config/locales/helpers/en.yml"}]
     # ./spec/i18n_spec.rb:16:in `block (2 levels) in <main>'
     # ./spec/rails_helper.rb:147:in `block (3 levels) in <top (required)>'
     # ./spec/rails_helper.rb:147:in `block (2 levels) in <top (required)>'
     # ./spec/rails_helper.rb:103:in `block (2 levels) in <top (required)>'
  2) I18n does not have inconsistent interpolations
     Failure/Error: expect(inconsistent_interpolations).to be_empty, error_message
     
       6 i18n keys have inconsistent interpolations.
       Run `i18n-tasks check-consistent-interpolations' to show them
     # ./spec/i18n_spec.rb:38:in `block (2 levels) in <main>'
     # ./spec/rails_helper.rb:147:in `block (3 levels) in <top (required)>'
     # ./spec/rails_helper.rb:147:in `block (2 levels) in <top (required)>'
     # ./spec/rails_helper.rb:103:in `block (2 levels) in <top (required)>'
Finished in 7 minutes 59 seconds (files took 7.37 seconds to load)
2360 examples, 2 failures, 19 pending
Failed examples:
rspec ./spec/i18n_spec.rb:15 # I18n does not have missing keys
rspec ./spec/i18n_spec.rb:35 # I18n does not have inconsistent interpolations
```

Seems like this is due to the typo Dan mentioned previously: `authentification_helper` should be `authentication_helper` in `config/locales/helpers/fr.yml`

Would you mind changing that @thomasbnt and hopefully we should see things start to pass?

thomasbnt commented:
@aitchiss Ooh yes of course! Edited !

aitchiss commented:
We're getting closer with the CI build - I can feel it!

I'm now seeing just this error:

```
Failures:
  1) I18n does not have inconsistent interpolations
     Failure/Error: expect(inconsistent_interpolations).to be_empty, error_message
     
       6 i18n keys have inconsistent interpolations.
       Run `i18n-tasks check-consistent-interpolations' to show them
     # ./spec/i18n_spec.rb:38:in `block (2 levels) in <main>'
     # ./spec/rails_helper.rb:147:in `block (3 levels) in <top (required)>'
     # ./spec/rails_helper.rb:147:in `block (2 levels) in <top (required)>'
     # ./spec/rails_helper.rb:103:in `block (2 levels) in <top (required)>'
```

I had a quick check and I think the problem is the parameter names passed in have _also_ been localised, when the parameter name should stay in English (since it's only used in code).

It seems to consistently be the parameter "community", and needs changed in the following locations (all instances in `%{}` I think should be "community"):

```
+--------+---------------------------------------------------+------------------------------------------------------------------+
| Locale | Key                                               | Value                                                            |
+--------+---------------------------------------------------+------------------------------------------------------------------+
|   fr   | helpers.feedback_messages_helper.affected.subject | Avis de courtoisie de %{communauté}                              |
|   fr   | json.author                                       | Auteur actif de la %{communauté}                                 |
|   fr   | mailers.digest_mailer.more_top_posts              | plus de meilleurs articles de %{communtiy}                       |
|   fr   | mailers.digest_mailer.more_top_posts_based        | plus de meilleurs articles de %{communtiy} basé sur vos intérêts |
|   fr   | mailers.digest_mailer.other_top_posts             | d'autres articles tendance de la %{communauté}                   |
|   fr   | mailers.digest_mailer.other_trending_posts        | d'autres articles tendance de la %{communauté}.                  |
+--------+---------------------------------------------------+------------------------------------------------------------------
```

msarit commented:
@thomasbnt This is a MONUMENTAL effort! Thank you so much!

Could we resolve the conflicts and continue addressing the feedback in the PR? Let us know if you need more support. 🙏🏾

thomasbnt commented:
> @thomasbnt This is a MONUMENTAL effort! Thank you so much!
> 
> Could we resolve the conflicts and continue addressing the feedback in the PR? Let us know if you need more support. 🙏🏾

The conflict is resolved 😄🚀

yheuhtozr commented:
@djuber @thomasbnt It's not a big deal but could you check [this comment of mine](https://github.com/forem/forem/pull/17487#discussion_r872070930) if you don't mind? There appears to be some confusion.

thomasbnt commented:
> @djuber @thomasbnt It's not a big deal but could you check [this comment of mine](https://github.com/forem/forem/pull/17487#discussion_r872070930) if you don't mind? There appears to be some confusion.

Oh hello @yheuhtozr, what is your comment? Sorry for late response, I didn't see your message 😶

yheuhtozr commented:
@thomasbnt Hi, if you see my reply you just quoted on GitHub UI, you can follow the link to jump to where I mentioned. It is all about a tiny discussion months ago. Just in case, the accurate place: https://github.com/forem/forem/pull/17487#discussion_r872070930

thomasbnt commented:
I didn't see your comment 😅 @yheuhtozr

thomasbnt commented:
Updated this PR with new file `config/locales/views/moderations/fr.yml`. 

I'm not sure about the translation of the word "Tag", if it should stay as such, or translated as "balise". I wrote "étiquette".

If some French people come here, don't hesitate to help me and to review the files. 👋🏼

jaw6 reviewed (changes requested):
@thomasbnt It looks like this is very, very close to passing the remaining I18n checks. I think with just a couple of tweaks we ought to have it!

thomasbnt commented:
Ow woah good catch @jaw6 !

jaw6 commented on config/locales/helpers/fr.yml:
```diff
@@ -4,229 +4,229 @@ fr:
-      author: Author
+      ask_me_anything: Demandez-moi n'importe quoi
+      author: Auteur
+      like: Aime
```
Seems this is no longer used after fd690982c -- easily missed when merging in from main!

```suggestion
```

thomasbnt commented:
Miss Line 117 to the latest edited file. Resolve a conflict on a mobile... It's difficult 😜

maestromac reviewed (approved):
Thank you so much for this tall task @thomasbnt !

maestromac commented:
I'm not sure why the specs are failing right now, but I am investigating.

thomasbnt commented:
> I'm not sure why the specs are failing right now, but I am investigating.
Thanks!

maestromac commented on config/locales/controllers/fr.yml:
```diff
@@ -1,123 +1,122 @@
     create:
-      success: created
-      failure: comment already exists
-      authorization_error: Your privileges have been suspended.
```
This key is missing. That's why the spec was failing.

thomasbnt replied:
Thanks @maestromac, I added `authorization_error` 👍🏼

maestromac reviewed (approved):
LGTM!
